packages/runtime/container-runtime/src/dataStores.ts

```diff
@@ -377,11 +379,11 @@ export class DataStores implements IDisposable {
         return entries;
     }
 
-    public async summarizeInternal(
-        builder: SummaryTreeBuilder, fullTree: boolean, trackState: boolean): Promise<void> {
+    public async summarize(fullTree: boolean, trackState: boolean): Promise<IChannelSummarizeResult> {
+        const builder = new SummaryTreeBuilder();
         // A list of this channel's GC nodes. Starts with this channel's GC node and adds the GC nodes all its child
         // channel contexts.
-        let gcNodes: IGraphNode[] = [ this.getGCNode() ];
+        let gcNodes: IGraphNode[] = [ await this.getGCNode() ];
```

Comment:
I worry that most of the code around GC assumes (without asserting) that document can't change. I.e. ops are paused, no external users can add / delete anything (data store / dds).
It would be great to sprinkle code with asserts in places where such assumption is the core for correctness. I worry if we need to change design (let's say in future get rid of summarizer client and teach main client to run GC in presence of local changes, without pausing ops), that it would be extremely hard to impossible to extract all assumptions in GC-related code and not to break them
Basically, defense in depth, fail fast.

Reply:
I agree. However today, summarize can be called from multiple places and I am not sure if all of them pause delta manager, etc. I believe this will need some thorough investigation before we can add such fail fasts.

I will add such asserts for GC in one of my subsequent PRs where I separate out GC and summarize.